File: backend/src/ips_app/ports/driving/http/role.py
from abc import ABC, abstractmethod
from typing import Optional, List, Tuple, Any
from ips_app.domain.models.role import Role
from ips_app.domain.models.permission import Permission

class RoleHTTPPort(ABC):
    @abstractmethod
    async def add_role(self, name: str, description: str) -> Role:
        """Add a new role."""
        ...

    @abstractmethod
    async def get_role(self, role_id: Any) -> Role:
        """Get a role by its ID."""
        ...

    @abstractmethod
    async def get_roles(
        self, 
        page: int, 
        limit: int, 
        cursor_id: Optional[Any] = None, 
        search: Optional[str] = None
    ) -> Tuple[List[Role], int]:
        """Get a list of roles with pagination."""
        ...

    @abstractmethod
    async def set_role(
        self, 
        role_id: Any, 
        name: Optional[str] = None, 
        description: Optional[str] = None
    ) -> Role:
        """Update a role's basic information."""
        ...

    @abstractmethod
    async def set_role_preferences(self, role_id: Any, preferences: bytes) -> Role:
        """Update a role's preferences."""
        ...

    @abstractmethod
    async def remove_role(self, role_id: Any) -> str:
        """Remove a role."""
        ...

    @abstractmethod
    async def add_permissions_to_role(self, role_id: Any, permission_ids: List[Any]) -> Role:
        """Bind permissions to a role."""
        ...

    @abstractmethod
    async def remove_permissions_from_role(self, role_id: Any, permission_ids: List[Any]) -> Role:
        """Unbind permissions from a role."""
        ...

    @abstractmethod
    async def get_role_permissions(self, role_id: Any) -> List[Permission]:
        """Get permissions bound to a role."""
        ...

File: backend/internal/ports/driving/http/role.go
package portsdrivinghttp

import (
	"context"

	domainmodels "github.com/uwb-ips-dteti-ugm/ips-app/backend/internal/domain/models"
)

type Role interface {
	// Roles
	AddRole(ctx context.Context, name string, description string) (role domainmodels.Role, err error)
	GetRole(ctx context.Context, roleId int64) (role domainmodels.Role, err error)
	GetRoles(ctx context.Context, page int, limit int, cursorId *int64, search *string) (roles []domainmodels.Role, total int, err error)
	SetRole(ctx context.Context, roleId int64, name *string, description *string) (role domainmodels.Role, err error)
	SetRolePreferences(ctx context.Context, roleId int64, preferences []byte) (role domainmodels.Role, err error)
	RemoveRole(ctx context.Context, roleId int64) (message string, err error)

	// Permissions
	AddPermission(ctx context.Context, name string, description string) (permission domainmodels.Permission, err error)
	GetPermission(ctx context.Context, permissionId int64) (permission domainmodels.Permission, err error)
	GetPermissions(ctx context.Context, page int, limit int, cursorId *int64, search *string) (permissions []domainmodels.Permission, total int, err error)
	SetPermission(ctx context.Context, permissionId int64, name *string, description *string) (permission domainmodels.Permission, err error)
	SetPermissionPreferences(ctx context.Context, permissionId int64, preferences []byte) (permission domainmodels.Permission, err error)
	RemovePermission(ctx context.Context, permissionId int64) (message string, err error)

	// Role-Permission bindings
	AddPermissionsToRole(ctx context.Context, roleId int64, permissionIds []int64) (role domainmodels.Role, err error)
	RemovePermissionsFromRole(ctx context.Context, roleId int64, permissionIds []int64) (role domainmodels.Role, err error)
	GetRolePermissions(ctx context.Context, roleId int64) (permissions []domainmodels.Permission, err error)
}
